File: typescript/tutorials/tour-of-workflows-typescript/src/workflows/signup-with-signals.ts
import * as restate from "@restatedev/restate-sdk";
import {
  WorkflowContext,
  WorkflowSharedContext,
} from "@restatedev/restate-sdk";
import { sendVerificationEmail, User } from "../utils";

export const signupWithSignals = restate.workflow({
  name: "SignupWithSignalsWorkflow",
  handlers: {
    run: async (ctx: WorkflowContext, user: User) => {
      const userId = ctx.key;

      // Generate verification secret and send email
      const secret = ctx.rand.uuidv4();
      await ctx.run("verify", () =>
        sendVerificationEmail(userId, user, secret),
      );

      // Wait for user to click verification link
      const clickedSecret = await ctx.promise<string>("email-verified");
      return { success: clickedSecret === secret };
    },
    verifyEmail: async (
      ctx: WorkflowSharedContext,
      req: { secret: string },
    ) => {
      // Resolve the promise to continue the main workflow
      await ctx.promise<string>("email-verified").resolve(req.secret);
    },
  },
});

File: python/tutorials/tour-of-workflows-python/app/workflows/signup_with_signals.py
import restate
from restate import WorkflowContext, WorkflowSharedContext
from app.utils import User, send_verification_email, VerifyEmailRequest

signup_with_signals = restate.Workflow("SignupWithSignalsWorkflow")


@signup_with_signals.main()
async def run(ctx: WorkflowContext, user: User) -> bool:
    user_id = ctx.key()

    # Generate verification secret and send email
    secret = str(ctx.uuid())
    await ctx.run_typed(
        "verify",
        send_verification_email,
        user_id=user_id,
        user=user,
        verification_secret=secret,
    )

    # Wait for user to click verification link
    clicked_secret = await ctx.promise("email-verified", type_hint=str).value()
    return clicked_secret == secret


@signup_with_signals.handler("verifyEmail")
async def verify_email(ctx: WorkflowSharedContext, req: VerifyEmailRequest) -> None:
    # Resolve the promise to continue the main workflow
    await ctx.promise("email-verified", type_hint=str).resolve(req.secret)
